### ISSUE TITLE
How to define model overfitting?

### ISSUE THREAD

**Issue by Shanginre** [signals-dev/Orion#160]
When model trains I get information for each epoch like

[Dx loss: [-3.9674070e-01 -1.4099512e+01  1.3591357e+01  1.1141491e-02]] 
[Dz loss: [-0.43088695  0.9507617  -1.8299156   0.04482671]] 
[G loss: [-11.348843  -13.889611    2.1240802   0.0416688]]

**how to find out from this information whether the overfitting of the model has begun or not?** 

As I understand from the pipeline, anomaly recognition occurs after the model has been fully trained after all epochs.

**Comment by sarahmish** [comment]
Hi @Shanginre, very good question!

The numbers presented depict the loss metric for each one of the neural networks used in the TadGAN model. You can find details on the architecture and components of TadGAN in the [blogpost](https://link.medium.com/cGsBD0Fevbb) or detailed in the [paper](https://arxiv.org/abs/2009.07769). The make takeaway is, there are three main models in the suggested GAN formulation
* The first discriminator (Dx) which is responsible for distinguishing between a real signal x and a fake signal.
* The second discriminator (Dz) which is responsible for distinguishing a random latent variable z and an encoded signal.
* A combination of a generator and an encoder (G) which is responsible for regenerating the encoded signal.

Let information shown for each model be captured as `loss[0] loss[1] loss[2] loss[3]` (you can see that the output is 4 float numbers). Then the loss shown above can be decomposed as:
```
loss[0] = loss[1] + loss[2] + 10 * loss[3]
```
For example, for Dz, `-0.43088695 = 0.9507617 -1.8299156 + 10 * 0.04482671`. Each one of the losses `loss[1] loss[2] loss[3]` represents a different loss function used to train the corresponding network. If you wish to investigate more about these parameters, I would visit the [tadgan](https://github.com/signals-dev/Orion/blob/master/orion/primitives/tadgan.py) implementation.

Now to tell if the model is overfitting or not, I would look at the mean square error metric. You will find this information within the last entry of G loss, in your case `0.0416688`. You can change some of the hyperparameters of the pipeline to fit your data's setting, you can also reduce/increase the number of epochs needed to verify your pipeline's performance.

I hope this answers your question :)

**Comment by Shanginre** [comment]
@sarahmish thanks for the answer! 

I conducted a series of numerical experiments and found a big problem. I used the benchmark function like this:

from orion.benchmark import benchmark
```
pipelines = ['lstm_dynamic_threshold']
signals = ['S-1']
datasets = {'demo': signals}
hyperparameters = {
     'demo': {
         'lstm_dynamic_threshold': {
             'keras.Sequential.LSTMTimeSeriesRegressor # 1': {'epochs': 35, 'verbose': True}
         }
     }
}
metrics = ['f1', 'accuracy', 'recall', 'precision']
scores = benchmark (pipelines = pipelines, datasets = datasets, hyperparameters = hyperparameters, metrics = metrics, rank = 'f1')
```
this is the output of the "scores" dataframe:
![image](https://user-images.githubusercontent.com/11506233/104873605-bacbe300-5972-11eb-90c4-d091fece1d30.png)

**but with the same parameters, another run of the benchmark function gave completely opposite results for the quality** of the "lstm_dynamic_threshold" model:
```
pipelines = ['lstm_dynamic_threshold', 'tadgan']
hyperparameters = {
    'demo_SP': {
        'lstm_dynamic_threshold': {
            'keras.Sequential.LSTMTimeSeriesRegressor # 1': {'epochs': 35, 'verbose': True}
        },
        'tadgan': {
            'orion.primitives.tadgan.TadGAN # 1': {'epochs': 50}
        }
    }
}
scores = benchmark (pipelines = pipelines, datasets = datasets, hyperparameters = hyperparameters, metrics = metrics, rank = 'f1')
```
this is the output of the "scores" dataframe (you can ignore the tadgan results):
![image](https://user-images.githubusercontent.com/11506233/104873745-05e5f600-5973-11eb-9aa2-c9228468a2d4.png)

**a similar "instability" is observed for the tadgan model**, and for any number of epochs (I tried different options).

Either I did not take into account some parameter or feature of the benchmark function, or there is a real problem in the proposed models.

**Comment by Shanginre** [comment]
@sarahmish I attach files with information on fittining lstm_dynamic_threshold:

https://drive.google.com/file/d/1DhFm5l7OWFkpwhsszJNkyy51QR9s-0By/view?usp=sharing
https://drive.google.com/file/d/1ehoDE8JJJYqvGWxbPFI3z6PTOp0MzBfX/view?usp=sharing

**Comment by sarahmish** [comment]
**TL;DR**

@Shanginre thank you for question. I ran the `lstm_dynamic_threshold` pipeline multiple times using the specifications you mentioned, and each time it produces the same results. 

**Details**
I used the specifications you listed earlier (mind I edited `signals=['S-1']`) and ran the benchmark (3 times); this is the result I obtained
```
In [1]: from orion.benchmark import benchmark 
   ...: pipelines = ['lstm_dynamic_threshold'] 
   ...: signals = ['P-1'] 
   ...: datasets = {'demo': signals} 
   ...: hyperparameters = { 
   ...:      'demo': { 
   ...:          'lstm_dynamic_threshold': { 
   ...:              'keras.Sequential.LSTMTimeSeriesRegressor#1': {'epochs': 35, 'verbose': True} 
   ...:          } 
   ...:      } 
   ...: } 
   ...: metrics = ['f1', 'accuracy', 'recall', 'precision'] 
   ...: scores = benchmark(pipelines=pipelines, datasets=datasets, hyperparameters=hyperparameters, metrics=metrics, rank='f1') 

In [2]: scores                                                                                                                                                                  
Out[2]: 
                 pipeline  rank  accuracy dataset      elapsed  f1  precision  recall signal  split status
0  lstm_dynamic_threshold     1  0.934248    demo  1503.066577 NaN        NaN     0.0    P-1  False     OK
```
In the other two runs, I also got 
```
                 pipeline  rank  accuracy dataset      elapsed    f1  precision  recall signal  split status
0  lstm_dynamic_threshold     1  0.934248    demo  1500.286987   NaN        NaN     0.0    P-1  False     OK
```
and with higher epoch got
```
                 pipeline  rank  accuracy dataset      elapsed  f1  precision  recall signal  split status
0  lstm_dynamic_threshold     1  0.929764    demo  3106.830503 NaN        0.0     0.0    P-1  False     OK
```
which to me look consistent and stable.
I'm wondering if there is another parameters you tweaked to cause the changes you see. For example, I can see that the model on top took significantly longer than what you see in your second run.. 

I also want to emphasize that we care about the stability and reproducibility of the models. We perform a benchmark in every release (which you can find the result of [here](https://docs.google.com/spreadsheets/d/1HaYDjY-BEXEObbi65fwG0om5d8kbRarhpK4mvOZVmqU/edit?usp=sharing)) and in this benchmark, you can also see consistency. Note that the scores used in benchmark evaluation is `weighted=False` which is different than the default ones you are using, you can easily modify it using:
```python3
from functools import partial
from orion.evaluation import contextual_f1_score, contextual_precision, contextual_recall

metrics = {
    "f1": partial(contextual_f1_score, weighted=False),
    "precision": partial(contextual_precision, weighted=False),
    "recall": partial(contextual_recall, weighted=False)
}
```
then pass the metric dictionary to the benchmark.
I also want to point that the _weighted segment_ approach (`weighted=True`) is very sensitive to the range of anomalies detected, which might be a reason for a slight change but not as significant as what you are showing.

With regards to `tadgan`, some variations are expected. Generally, GAN models are known to produce different results in different instantiations since the model parameters are actually randomly initialized. Based on the adversarial training, it might result in different scores in each run.

Hope this clarifies it for you.

**Comment by Shanginre** [comment]
@sarahmish thanks for the answer. 

This does clear things up a bit. I will try a series of experiments with "weighted = False".

The only thing that confuses me is the **NaN and 0** values for f1, precision recall in your results. As I understand the meaning of these parameters, such values of NaN and 0 indicate poor quality of model prediction on real data. 

Is this so and how to interpret such results?

**Comment by Shanginre** [comment]
I started testing "weighted = False" and **always get 0 for parameters 'f1', 'recall', 'precision'**
```
     metrics = {
         "f1": partial (contextual_f1_score, weighted = False),
         "accuracy": partial (contextual_accuracy, weighted = True),
         "precision": partial (contextual_precision, weighted = False),
         "recall": partial (contextual_recall, weighted = False)
     }
```
why is this happening and how to fix it? Can you run a series of experiments to test it?

**Comment by sarahmish** [comment]
The reason you see these scores is the fact that the pipeline did not manage to find any true positives (tp) in the signal it's trying to detect anomalies in. Remember how we compute the scores for each one of these methods:
```
precision = tp / (tp + fp)
recall = tp / (tp + fn)
f1_score = 2 * (precision * recall) / (precision + recall)
``` 
If you did not manage to detect any true positives (`tp = 0`) then by calculation, you would have
```
precision = 0 / (0 + fp) = 0
recall = 0 / (0 + fn) = 0
f1_score = 2 * (0 * 0) / (0 + 0) = NaN
``` 
This is the case for the signal you are experimenting with `P-1`, if you experiment with another signal, in this case I am showing `S-1`, you will get a different score
```
In [1]: from functools import partial 
   ...: from orion.evaluation import contextual_f1_score, contextual_precision, contextual_recall 
   ...:  
   ...: metrics = { 
   ...:     "f1": partial(contextual_f1_score, weighted=False), 
   ...:     "precision": partial(contextual_precision, weighted=False), 
   ...:     "recall": partial(contextual_recall, weighted=False) 
   ...: }   

In [2]: from orion.benchmark import benchmark
   ...: 
   ...: pipelines = ['lstm_dynamic_threshold'] 
   ...: signals = ['P-1', 'S-1'] 
   ...: datasets = {'demo': signals} 
   ...: hyperparameters = { 
   ...:      'demo': { 
   ...:          'lstm_dynamic_threshold': { 
   ...:              'keras.Sequential.LSTMTimeSeriesRegressor#1': {'epochs': 5, 'verbose': True} 
   ...:          } 
   ...:      } 
   ...: } 
   ...:  scores = benchmark(pipelines=pipelines,
   ...:      datasets=datasets, hyperparameters=hyperparameters, metrics=metrics, rank='f1')

In [3]: scores
Out[3]: 
                 pipeline  rank dataset     elapsed        f1  precision  recall signal  split status
0  lstm_dynamic_threshold     1    demo  294.671088  0.666667        0.5     1.0    S-1  False     OK
1  lstm_dynamic_threshold     2    demo  327.840807       NaN        NaN     0.0    P-1  False     OK

```

**Note**: I would like to note that `weighted=False` does not have an accuracy score since overlapping segment approach does not count the number of true negatives (tn) which is necessary for an accuracy score.

**Comment by Shanginre** [comment]
@sarahmish I have conducted a series of experiments. Run this script:
```
    pipelines = ['lstm_dynamic_threshold']
    signals = ['P-1']
    datasets = {'demo': signals}
    hyperparameters = {
        'demo': {
            'lstm_dynamic_threshold': {
                'keras.Sequential.LSTMTimeSeriesRegressor#1': {'epochs': 35, 'verbose': True}
            }
        }
    }
    metrics = ['accuracy', 'f1', 'recall', 'precision']

    is_first = True
    for i in range(5):
        print(i)
        if is_first:
            is_first = False
            results_df = benchmark(pipelines=pipelines, datasets=datasets, hyperparameters=hyperparameters, metrics=metrics, rank='f1')
        else:
            scores = benchmark(pipelines=pipelines, datasets=datasets, hyperparameters=hyperparameters, metrics=metrics, rank='f1')
            results_df = pd.concat([results_df, scores])
        print(results_df[['accuracy', 'f1', 'recall', 'precision']])
```
the results are as follows:
```
   accuracy        f1    recall  precision
0  0.907788  0.874656  0.785745   0.986257
0  0.612518       NaN  0.000000        NaN
0  0.612518       NaN  0.000000        NaN
0  0.612518       NaN  0.000000        NaN
0  0.612518       NaN  0.000000        NaN
0  0.907788  0.874656  0.785745   0.986257
0  0.612518       NaN  0.000000        NaN
0  0.907788  0.874656  0.785745   0.986257
0  0.907788  0.874656  0.785745   0.986257
0  0.612518       NaN  0.000000        NaN
```
That is, the model either shows good results or shows extremely poor results. **This fact is very disappointing. What can we do to make the quality of the model more sustainable?**

**Comment by Shanginre** [comment]
@sarahmish (addition to the question) for metrics
```
    metrics = {
        "f1": partial(contextual_f1_score, weighted=False),
        "accuracy": partial(contextual_accuracy, weighted=True),
        "precision": partial(contextual_precision, weighted=False),
        "recall": partial(contextual_recall, weighted=False)
    }
```
here are results:
```
   accuracy  f1  recall  precision
0  0.907788 NaN     0.0        0.0
0  0.612518 NaN     0.0        NaN
0  0.612518 NaN     0.0        NaN
0  0.907788 NaN     0.0        0.0
```
that is even worse. All parameters except accuracy have not been calculated. Although there are some values for their "weighted" variants. Similarly, the model either gives good accuracy 0.907788 or very poor accuracy 0.612518.

my last question is relevant: 
**how to make the quality of the model more stable?** 
After all, if the quality of the model is not stable, this makes your library not suitable for the IOT case (my case), where there are several hundred time series and there is no way to manually adjust (more like witchcraft) the model for each time series.

**Comment by Shanginre** [comment]
@sarahmish should i wait for a reply to my last message?

I want to use your library in my ETL pipeline, but so far its instability stops me.

**Comment by sarahmish** [comment]
Hi @Shanginre, I have a couple of thoughts here

**P-1 anomalies**
The signal you are testing on has three anomalies, namely:
```
        start         end
0  1331272800  1335592800
1  1382832000  1389484800
2  1361296800  1366480800
```
I suspected that because of stochasticity you're sometimes detecting an anomaly and sometimes missing it, this would explain why in some cases you got a better result than others (even if it's a single anomaly).

**Reproducing**
I ran the script you described above and got the following results
```
   accuracy  f1  recall  precision
0  0.934248 NaN     0.0        NaN
0  0.934248 NaN     0.0        NaN
0  0.934248 NaN     0.0        NaN
0  0.934248 NaN     0.0        NaN
0  0.934248 NaN     0.0        NaN
```
They are different to yours. I am wondering if you can create a google colab notebook, run the script that's causing an issue, and share it with me here such that I can figure out what's causing the change.

**Comment by Shanginre** [comment]
@sarahmish ок. I'll make a notebook.

what model parameters did you use in the test above? What version of Orion?

**Comment by sarahmish** [comment]
The latest release of Orion (version 0.1.5). I used the script you mentioned previously

> ```python3
> import pandas as pd
> from orion.benchmark import benchmark
> 
> pipelines = ['lstm_dynamic_threshold']
> signals = ['P-1']
> datasets = {'demo': signals}
> hyperparameters = {
>     'demo': {
>         'lstm_dynamic_threshold': {
>             'keras.Sequential.LSTMTimeSeriesRegressor#1': {'epochs': 35, 'verbose': True}
>         }
>     }
> }
> 
> metrics = ['accuracy', 'f1', 'recall', 'precision']
> 
> is_first = True
> for i in range(5):
>     print(i)
>     if is_first:
>         is_first = False
>         results_df = benchmark(pipelines=pipelines, datasets=datasets, hyperparameters=hyperparameters, metrics=metrics, rank='f1')
>     else:
>         scores = benchmark(pipelines=pipelines, datasets=datasets, hyperparameters=hyperparameters, metrics=metrics, rank='f1')
>         results_df = pd.concat([results_df, scores])
> 
> print(results_df[['accuracy', 'f1', 'recall', 'precision']])
> ```
